nucliadb/nucliadb/common/datamanagers/rollover.py

```diff
@@ -0,0 +1,151 @@
+
+    async def add_batch_to_index(self, kbid: str, batch: list[str]) -> None:
+        async with self.driver.transaction() as txn:
+            for key in batch:
+                key = KB_ROLLOVER_RESOURCES_TO_INDEX.format(kbid=kbid, resource=key)
+                await txn.set(key, b"")
+            await txn.commit()
+
+    async def get_to_index(self, kbid: str) -> Optional[str]:
+        key = KB_ROLLOVER_RESOURCES_TO_INDEX.format(kbid=kbid, resource="")
+        async with self.driver.transaction(wait_for_abort=False) as txn:
+            found = [key async for key in txn.keys(key, count=1)]
+            if found:
+                return found[0].split("/")[-1]
+        return None
+
+    async def remove_to_index(self, kbid: str, resource: str) -> None:
+        to_index = KB_ROLLOVER_RESOURCES_TO_INDEX.format(kbid=kbid, resource=resource)
+        async with self.driver.transaction() as txn:
+            await txn.delete(to_index)
+            await txn.commit()
+
+    async def add_indexed(
+        self, kbid: str, resource_id: str, shard_id: str, modification_time: int
+    ) -> None:
+        to_index = KB_ROLLOVER_RESOURCES_TO_INDEX.format(
+            kbid=kbid, resource=resource_id
+        )
+        indexed = KB_ROLLOVER_RESOURCES_INDEXED.format(kbid=kbid, resource=resource_id)
+        data = [shard_id, modification_time]
+        async with self.driver.transaction() as txn:
+            await txn.delete(to_index)
+            await txn.set(indexed, orjson.dumps(data))
+            await txn.commit()
+
+    async def get_indexed_data(
+        self, kbid: str, resource_id: str
+    ) -> Optional[tuple[str, int]]:
+        key = KB_ROLLOVER_RESOURCES_INDEXED.format(kbid=kbid, resource=resource_id)
+        async with self.driver.transaction(wait_for_abort=False) as txn:
+            val = await txn.get(key)
+            if val is not None:
+                data = orjson.loads(val)
+                return tuple(data)  # type: ignore
+        return None
+
+    async def remove_indexed(self, kbid: str, batch: list[str]) -> None:
+        async with self.driver.transaction(wait_for_abort=False) as txn:
+            for resource_id in batch:
+                key = KB_ROLLOVER_RESOURCES_INDEXED.format(
+                    kbid=kbid, resource=resource_id
+                )
+                await txn.delete(key)
+            await txn.commit()
+
+    async def iter_indexed_keys(self, kbid: str) -> AsyncGenerator[str, None]:
+        start_key = KB_ROLLOVER_RESOURCES_INDEXED.format(kbid=kbid, resource="")
+        async with self.driver.transaction(wait_for_abort=False) as txn:
+            async for key in txn.keys(match=start_key, count=-1):
+                yield key.split("/")[-1]
+
+    async def _get_batch_indexed_data(
+        self, kbid, batch: list[str]
+    ) -> list[tuple[str, tuple[str, int]]]:
+        async with self.driver.transaction(wait_for_abort=False) as txn:
+            values = await txn.batch_get(
```

Comment:
I wonder if this batch get can be problematic? Sometimes a single index message can be too big to fit into the max grpc payload of tikv, no?

Reply:
ah, nevermind. I see it now. I thought you were storing the whole brain message in tikv. sorry